README.md
# tslint-rules

[![Greenkeeper badge](https://badges.greenkeeper.io/Alorel/tslint-rules.svg)](https://greenkeeper.io/)

Personal TSLint rule presets


CHANGELOG.md
## [5.0.7](https://github.com/Alorel/tslint-rules/compare/5.0.6...5.0.7) (2020-04-18)


### Bug Fixes

* Update some rules to be less annoying ([925cf24](https://github.com/Alorel/tslint-rules/commit/925cf24))


### Maintenance

* **deps:** bump npm from 6.10.1 to 6.13.4 ([9862d62](https://github.com/Alorel/tslint-rules/commit/9862d62))

## [5.0.6](https://github.com/Alorel/tslint-rules/compare/5.0.5...5.0.6) (2019-11-13)


### Bug Fixes

* Remove completed-docs from common.json until they fix their error spam ([2b60354](https://github.com/Alorel/tslint-rules/commit/2b60354))

## [5.0.5](https://github.com/Alorel/tslint-rules/compare/5.0.4...5.0.5) (2019-07-17)


### Bug Fixes

* Remove no-non-null-assertion rule ([8f9d643](https://github.com/Alorel/tslint-rules/commit/8f9d643))

## [5.0.4](https://github.com/Alorel/tslint-rules/compare/5.0.3...5.0.4) (2019-07-16)


### Bug Fixes

* Reverted commit 1d8f8035 ([cc49a7b](https://github.com/Alorel/tslint-rules/commit/cc49a7b))

## [5.0.3](https://github.com/Alorel/tslint-rules/compare/5.0.2...5.0.3) (2019-07-16)


### Bug Fixes

* rm one-variable-per-declaration ([f76c0b2](https://github.com/Alorel/tslint-rules/commit/f76c0b2))
* Use ^ version range for tslint-sonarts ([0ee48ed](https://github.com/Alorel/tslint-rules/commit/0ee48ed)), closes [#15](https://github.com/Alorel/tslint-rules/issues/15)
* Use array-simple type ([1d8f803](https://github.com/Alorel/tslint-rules/commit/1d8f803))

## [5.0.2](https://github.com/Alorel/tslint-rules/compare/5.0.1...5.0.2) (2019-07-15)


### Bug Fixes

* add tags/existence to completed-docs ([42be9e5](https://github.com/Alorel/tslint-rules/commit/42be9e5))

## [5.0.1](https://github.com/Alorel/tslint-rules/compare/5.0.0...5.0.1) (2019-07-15)


### Bug Fixes

* Remove strict-comparisons rule ([9d43702](https://github.com/Alorel/tslint-rules/commit/9d43702))

# [5.0.0](https://github.com/Alorel/tslint-rules/compare/4.1.1...5.0.0) (2019-07-15)


### Dependency updates

* **package:** update tslint to version 5.18.0 ([e67c405](https://github.com/Alorel/tslint-rules/commit/e67c405))
* **package:** update tslint-sonarts to version 1.9.0 ([6af5840](https://github.com/Alorel/tslint-rules/commit/6af5840))


### Features

* Add SonarTS 1.9.0 rules ([26d669c](https://github.com/Alorel/tslint-rules/commit/26d669c))
* Update standard tslint rules with additions between tslint 5.11.0 and 5.18.0 ([4af6bbb](https://github.com/Alorel/tslint-rules/commit/4af6bbb))


### Maintenance

* **deps:** bump fstream from 1.0.11 to 1.0.12 ([430f4c3](https://github.com/Alorel/tslint-rules/commit/430f4c3))
* **deps:** bump handlebars from 4.0.12 to 4.1.2 ([d762a2d](https://github.com/Alorel/tslint-rules/commit/d762a2d))
* **deps:** bump js-yaml from 3.12.0 to 3.13.1 ([fc87721](https://github.com/Alorel/tslint-rules/commit/fc87721))
* **deps:** bump lodash from 4.17.11 to 4.17.14 ([2fbc2c9](https://github.com/Alorel/tslint-rules/commit/2fbc2c9))


### BREAKING CHANGES

* Tslint rules have changed
* New rules from SonarTS added

## [4.1.1](https://github.com/Alorel/tslint-rules/compare/4.1.0...4.1.1) (2018-11-10)


### Bug Fixes

* Add 60000 to allowed magic numbers ([c87dea5](https://github.com/Alorel/tslint-rules/commit/c87dea5))

# [4.1.0](https://github.com/Alorel/tslint-rules/compare/4.0.0...4.1.0) (2018-10-26)


### Features

* Add the following allowed magic numbers: 999, 59, 23, 3600, 3600000, 86400, 86400000, 28, 30, 31 ([2ff406b](https://github.com/Alorel/tslint-rules/commit/2ff406b))

# [4.0.0](https://github.com/Alorel/tslint-rules/compare/3.0.0...4.0.0) (2018-10-09)


### Features

* Add tslint-sonarts ruleset. ([ac564a4](https://github.com/Alorel/tslint-rules/commit/ac564a4))


### BREAKING CHANGES

* A plethora of new rules have been introduced in sonarts.json and are likely to break existing applications.

# [3.0.0](https://github.com/Alorel/tslint-rules/compare/2.0.0...3.0.0) (2018-09-26)


### Features

* **config:** Add prefer-readonly rule ([c85eb68](https://github.com/Alorel/tslint-rules/commit/c85eb68))


### BREAKING CHANGES

* **config:** prefer-readonly now set to true

# [2.0.0](https://github.com/Alorel/tslint-rules/compare/1.2.2...2.0.0) (2018-09-04)


### Bug Fixes

* Allow spaces before function parens @ async arrow & anonymous ([3c8fdb9](https://github.com/Alorel/tslint-rules/commit/3c8fdb9))


### BREAKING CHANGES

* may require a tslint --fix

## [1.2.2](https://github.com/Alorel/tslint-rules/compare/1.2.1...1.2.2) (2018-08-17)


### Bug Fixes

* **package:** Add yarn.lock to npmignore ([b3a428e](https://github.com/Alorel/tslint-rules/commit/b3a428e))


### Maintenance

* **package:** Move release config to separate file ([f46e1da](https://github.com/Alorel/tslint-rules/commit/f46e1da))
* **package:** update [@alorel-personal](https://github.com/alorel-personal)/conventional-changelog-alorel to version 2.0.0 ([4e76d9a](https://github.com/Alorel/tslint-rules/commit/4e76d9a))
* **package:** update [@alorel-personal](https://github.com/alorel-personal)/conventional-changelog-alorel to version 2.0.1 ([1b0192a](https://github.com/Alorel/tslint-rules/commit/1b0192a))

## [1.2.1](https://github.com/Alorel/tslint-rules/compare/1.2.0...1.2.1) (2018-08-09)


### Bug Fixes

* Add missing 100 to no-magic-numbers ([56e8f7a](https://github.com/Alorel/tslint-rules/commit/56e8f7a))

# [1.2.0](https://github.com/Alorel/tslint-rules/compare/1.1.0...1.2.0) (2018-08-09)


### Features

* add more magic numbers to the whitelist ([a8a0c61](https://github.com/Alorel/tslint-rules/commit/a8a0c61))


### Maintenance

* **package:** Lock down conventional-changelog-alorel version ([0cb50b9](https://github.com/Alorel/tslint-rules/commit/0cb50b9))
* **package:** Refresh package-lock.json ([edb97fa](https://github.com/Alorel/tslint-rules/commit/edb97fa))

test.js
const fs = require('fs');
const {strictEqual} = require('assert');
const {join} = require('path');

strictEqual(fs.existsSync(join(__dirname, 'common.json')), true, 'common.json missing');
strictEqual(fs.existsSync(join(__dirname, 'sonarts.json')), true, 'sonarts.json missing');
strictEqual(fs.existsSync(join(__dirname, 'tslint.json')), true, 'tslint.json missing');

const extd = require('./tslint.json').extends;

strictEqual(extd.includes('./sonarts.json'), true, 'sonarts.json not extended');
strictEqual(extd.includes('./common.json'), true, 'common.json not extended');


generate.js
const fs = require('fs');
const {join} = require('path');

const promiseTypes = [
  'Bluebird',
  'Thenable',
  'Query' // mongoose
];

const rules = {
  'adjacent-overload-signatures': true,
  'ban-comma-operator': true,
  'import-blacklist': [
    true,
    'lodash'
  ],
  'member-ordering': [
    true,
    {
      'order': [
        'public-static-field',
        'protected-static-field',
        'private-static-field',
        'public-instance-field',
        'protected-instance-field',
        'private-instance-field',
        'public-constructor',
        'protected-constructor',
        'private-constructor',
        'public-static-method',
        'protected-static-method',
        'private-static-method',
        'public-instance-method',
        'protected-instance-method',
        'private-instance-method'
      ]
    }
  ],
  'no-empty-interface': true,
  'no-inferrable-types': true,
  'member-access': [
    true,
    'check-accessor',
    'check-constructor',
    'check-parameter-property'
  ],
  'no-magic-numbers': [
    true,
    -1,
    0,
    1,
    10,
    100,
    999,
    1000,
    59,
    60,
    60000,
    23,
    24,
    3600,
    3600000,
    86400,
    86400000,
    28,
    30,
    31,
    365,
    1024,
    2048,
    4096,
    8192,
    16384,
    32768,
    65536
  ],
  'no-namespace': false,
  'no-non-null-assertion': false,
  'no-reference': true,
  'no-unnecessary-type-assertion': true,
  'no-var-requires': true,
  'unified-signatures': true,
  'await-promise': [
    true,
    ...promiseTypes
  ],
  curly: [
    true,
    'ignore-same-line'
  ],
  forin: true,
  'label-position': true,
  'no-arg': true,
  'no-bitwise': true,
  'no-construct': true,
  'no-duplicate-super': true,
  'no-debugger': true,
  'no-duplicate-switch-case': true,
  'no-duplicate-variable': [
    true,
    'check-parameters'
  ],
  'no-empty': [
    true,
    'allow-empty-catch'
  ],
  'no-eval': true,
  'no-floating-promises': [
    true,
    ...promiseTypes
  ],
  'no-for-in-array': true,
  'no-implicit-dependencies': [
    true,
    'dev'
  ],
  'no-inferred-empty-object-type': true,
  'no-invalid-template-strings': true,
  'no-invalid-this': [
    true,
    'check-function-in-method'
  ],
  'no-misused-new': true,
  'no-object-literal-type-assertion': true,
  'no-return-await': true,
  'no-shadowed-variable': [
    true,
    {
      class: true,
      enum: true,
      function: true,
      import: true,
      interface: true,
      namespace: true,
      typeAlias: true,
      typeParameter: true,
      underscore: false
    }
  ],
  'no-sparse-arrays': true,
  'no-string-throw': true,
  'no-switch-case-fall-through': true,
  'no-this-assignment': [
    true,
    {
      'allow-destructuring': true
    }
  ],
  'no-unbound-method': [
    true,
    'allow-delete',
    'allow-typeof'
  ],
  'no-unsafe-finally': true,
  'no-unused-expression': [
    true,
    'allow-fast-null-checks',
    'allow-new'
  ],
  'no-var-keyword': true,
  'no-void-expression': [
    true,
    'ignore-arrow-function-shorthand'
  ],
  'prefer-conditional-expression': [
    true,
    'check-else-if'
  ],
  radix: false,
  'restrict-plus-operands': true,
  'triple-equals': [
      true,
      'allow-null-check',
      'allow-undefined-check'
  ],
  'use-isnan': true,
  'cyclomatic-complexity': [
    true,
    20
  ],
  deprecation: true,
  eofline: true,
  indent: [
    true,
    'spaces',
    2
  ],
  'linebreak-style': [
    true,
    'LF'
  ],
  'max-classes-per-file': [
    true,
    1
  ],
  'max-file-line-count': [
    true,
    300
  ],
  'max-line-length': [
    true,
    120
  ],
  'no-default-export': true,
  'no-duplicate-imports': true,
  'object-literal-sort-keys': [
    true,
    'ignore-case'
  ],
  align: [
    true,
    'parameters',
    'arguments',
    'statements',
    'members',
    'elements'
  ],
  'array-type': [
    true,
    'array'
  ],
  'arrow-parens': [
    true,
    'ban-single-arg-parens'
  ],
  'arrow-return-shorthand': true,
  'binary-expression-operand-order': true,
  'callable-types': true,
  'class-name': true,
  encoding: true,
  'import-spacing': true,
  'interface-over-type-literal': true,
  'newline-before-return': true,
  'new-parens': true,
  'no-consecutive-blank-lines': [
    true,
    1
  ],
  'no-irregular-whitespace': true,
  'no-reference-import': true,
  'no-trailing-whitespace': [
    true,
    'ignore-template-strings'
  ],
  'no-unnecessary-initializer': true,
  'no-unnecessary-qualifier': true,
  'number-literal-format': true,
  'object-literal-key-quotes': [
    true,
    'as-needed'
  ],
  'one-line': [
    true,
    'check-catch',
    'check-finally',
    'check-else',
    'check-open-brace',
    'check-whitespace'
  ],
  'one-variable-per-declaration': false,
  'ordered-imports': [
    true,
    {
      'import-sources-order': 'case-insensitive',
      'named-imports-order': 'case-insensitive'
    }
  ],
  'prefer-method-signature': true,
  'prefer-switch': true,
  'prefer-template': [
    true,
    'allow-single-concat'
  ],
  quotemark: [
    true,
    'single',
    'avoid-escape'
  ],
  semicolon: [
    true,
    'always',
    'strict-bound-class-methods'
  ],
  'space-before-function-paren': [
    true,
    {
      anonymous: 'always',
      named: 'never',
      asyncArrow: 'always',
      method: 'never',
      constructor: 'never'
    }
  ],
  'space-within-parens': [
    true,
    0
  ],
  'prefer-readonly': true,
  'switch-final-break': true,
  'type-literal-delimiter': [true, 'always'],
  'variable-name': [
    true,
    'check-format',
    'allow-leading-underscore',
    'require-const-for-all-caps',
    'ban-keywords'
  ],
  'return-undefined': true,
  'file-name-casing': false,
  'strict-boolean-expressions': false,
  'ban-ts-ignore': false,
  'function-constructor': true,
  'increment-decrement': false,
  'no-default-import': false,
  'unnecessary-constructor': true,
  'comment-type': false,
  'unnecessary-bind': true,
  'no-restricted-globals': false,
  'completed-docs': false,
  // 'completed-docs': [
  //   true,
  //   (() => {
  //     const exported = () => ({visibilities: ['exported']});
  //     const privacies = () => ({privacies: ['protected', 'public']});
  //     const tags = () => ({
  //       tags: {
  //         existence: [
  //           'internal',
  //           'inheritDoc',
  //           'see'
  //         ]
  //       }
  //     });
  //     const exptags = () => ({
  //       ...exported(),
  //       ...tags()
  //     });
  //
  //     return {
  //       classes: exptags(),
  //       enums: exptags(),
  //       'enum-members': false,
  //       functions: {
  //         overloads: true,
  //         ...exptags()
  //       },
  //       interfaces: exptags(),
  //       methods: {
  //         locations: 'all',
  //         ...privacies(),
  //         ...exptags()
  //       },
  //       namespaces: false,
  //       properties: {
  //         locations: 'all',
  //         ...privacies(),
  //         ...tags()
  //       },
  //       types: exptags(),
  //       variables: exptags()
  //     };
  //   })()
  // ],
  'no-tautology-expression': true,
  'static-this': false,
  'unnecessary-else': false,
  'no-any': false,
  'no-null-undefined-union': false,
  'no-async-without-await': true,
  'strict-comparisons': false,
  whitespace: [
    true,
    'check-branch',
    'check-decl',
    'check-operator',
    'check-separator',
    'check-rest-spread',
    'check type',
    'check-type-operator',
    'check-preblock'
  ]
};

fs.writeFileSync(join(__dirname, 'common.json'), JSON.stringify({rules}, null, 2));
